src/utmpx.rs

```diff
@@ -0,0 +1,234 @@
+            ut_id: Self::charbuf_to_bytes(&entry.ut_id),
+            ut_line: Self::charbuf_to_string(&entry.ut_line)?,
+            ut_pid: Pid::from_raw(entry.ut_pid),
+            ut_tv: TimeVal::from_usec(entry.ut_tv.tv_sec.into(), entry.ut_tv.tv_usec.into()),
+            ut_type: EntryType::try_from(entry.ut_type)?,
+            ut_user: Self::charbuf_to_string(&entry.ut_user)?,
+        };
+
+        Ok(res)
+    }
+
+    /// Copy out buffer content, as a bytes vector.
+    fn charbuf_to_bytes(input: &[libc::c_char]) -> Vec<u8> {
+        let bytes = unsafe { &*(input as *const _  as *const [u8]) };
+        let mut buf = vec![0u8; bytes.len()];
+        buf.copy_from_slice(bytes);
+        buf
+    }
+
+    /// Copy out buffer content, as a string, up to the first NUL-terminator.
+    fn charbuf_to_string(input: &[libc::c_char]) -> Result<String> {
+        // Ensure input is NUL-terminated (`ut_id` often isn't).
+        let mut buf: Vec<_> = input.to_vec();
+        buf.push(0);
+
+        let wrapped = unsafe { CStr::from_ptr(buf.as_ptr()) };
+        wrapped
+            .to_str()
+            .map(|s| s.to_string())
+            .map_err(|_| Error::InvalidUtf8)
+    }
+}
+
+/// Iterator over accounting entries.
+///
+/// Created using `Utmp.entries()`.
+#[derive(PartialEq, Eq, Debug)]
+pub struct UtmpIter<'a> {
+    db: &'a mut Utmp,
+}
+
+impl<'a> Iterator for UtmpIter<'a> {
+    type Item = Result<UtmpEntry>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = unsafe { libc::getutxent() };
+        // Stop iterating on listing error.
+        let entry_ptr = match Errno::result(res) {
+            Ok(ptr) => ptr,
+            _ => return None,
+        };
+        Some(UtmpEntry::try_from_ptr(entry_ptr))
+    }
+}
+
+/// Utmp accounting database.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Utmp {
+    /// Mark the DB and iterator as non-Send and non-Sync.
+    _phantom: PhantomData<*mut ()>,
+}
+
+impl Utmp {
+    /// Open the default `utmp` database, positioning pointer at the beginning.
+    ///
+    /// # Safety
+    ///
+    /// This operations is unsafe because it mutates global libc state. In order to
+    /// safely invoke this, the caller must ensure that nothing else in the process
+    /// is accessing the `utmp` database at the same time.
+    pub unsafe fn open() -> Result<Utmp> {
+        let mut db = Utmp {
+            _phantom: PhantomData,
+        };
+        db.rewind();
+        Ok(db)
+    }
+
+    /// Iterate through accounting entries.
+    pub fn entries(&mut self) -> UtmpIter {
+        UtmpIter { db: self }
+    }
+
+    /// Rewind the file pointer to the beginning of the database.
+    ///
+    /// # Safety
+    ///
+    /// This operations is unsafe because it mutates global libc state. In order to
+    /// safely invoke this, the caller must ensure that nothing else in the process
+    /// is accessing the `utmp` database at the same time.
+    pub unsafe fn rewind(&mut self) {
```

Comment:
I think this method can be safe, right?  Because it requires a `Utmp` instance, it can't be invoked without the caller already doing something `unsafe`.  And the `&mut` reference will prevent it from being invoked while there is an active `UtmpIter`.  The only problem I can foresee would be if the caller created two `Utmp` instances, but that would violate the rules for `Utmp::open`.